### ISSUE TITLE
Implement Hermes commands and eBPF processing

### ISSUE THREAD

**Issue by iomartin** [Eideticom/eid-hermes-qemu#8]
Now that the Hermes QEMU device can do XDMA, the next step is to implement the commands defined on this [eid-hermes PR](https://github.com/Eideticom/eid-hermes/pull/8).

This entails the following:
- Defining [BAR0 registers](https://github.com/Eideticom/eid-hermes/blob/master/specs/eid-hermes-interface.md) to receive the command requests
- Adding the QEMU logic to execute the 0x00, 0x01 and 0x80 commands. Command 0x80 should use [uBPF](https://github.com/iovisor/ubpf).
- Sending an interrupt back to the host when the command has completed.

Note that while we don't have a Hermes driver, we can exercise BAR0 with [pcimem](https://github.com/billfarrow/pcimem) (which is installed by the [ansible playbook](https://github.com/Eideticom/eid-hermes/tree/master/ansible)).

**Comment by niclashedam** [comment]
I'll give this a go as well. I would like a few pointers on how to start, like the other issue (#9). Thanks!

**Comment by iomartin** [comment]
Hey @niclashedam,

- I'd start by going through the the [specs](https://github.com/Eideticom/eid-hermes/tree/martin/driver_v3/specs) directory of eid-hermes, in particular eid-hermes-commands-format.md and eid-hermes-interface.md. The former defines the Hermes commands requests/responses and the latter defines the registers locations on BAR0 (the latest version of these docs are currently in the martin/driver_v3 branch, they should be merged into master in the coming days).

- You'll need to add those registers to the QEMU device. We don't have the driver code yet to generate the command requests, but you can manually write to those registers using the [pcimem](https://github.com/billfarrow/pcimem) program. For example, to write value 0xABCD to address 0x1000 of BAR0:

> $ sudo pcimem /sys/devices/pci0000\:00/0000\:00\:04.0/resource0 0x1000 w 0xABCD

It should already be installed on the VM. See the pcimem git repo and the tool help for more options.

- From the uBPF project, this is an example of how to use the uBPF VM: https://github.com/iovisor/ubpf/blob/master/vm/test.c

- Last year, I did an RFC eBPF device, it was kind of clunky and used a completely different API. It can serve as a starting point, but don't feel too tied to it: https://github.com/Eideticom/eid-hermes-qemu/blob/martin/bpf/hw/misc/pcie-ebpf.c

- We'll want to have multiple eBPF engines that can run in parallel

- Initially you can ignore the request/release slots commands (0x00/0x01). The hermes driver that I wrote can already do something equivalent in software. You can focus on command 0x80 for now.

- For the interrupts, see how we do it on the XDMA path.

**Comment by iomartin** [comment]
@niclashedam I created a draft PR with driver support for executing programs (https://github.com/Eideticom/eid-hermes/pull/23). I'd appreciate it if you could give it a try and review it once you're further along with the eBPF engines.
